File: solutions_2453486_1/Python/aquamongoose/tictactomek.py
#!/usr/bin/env python3
# Bill Cooperman (aquamongoose)
# Google Code Jam Qualification Round
# Problem A
T = int(input())
N = 4
board = ['.'*N]*N

def going():
    global board, N
    for line in board:
        if line.find('.') != -1:
            return True
    return False

def win(p):
    global board, N
    for i in range(N):
        g1 = g2 = True
        for j in range(N):
            if board[i][j] not in {p, 'T'}:
                g1 = False
            if board[j][i] not in {p, 'T'}:
                g2 = False
        if g1 or g2:
            return True
    g1 = g2 = True
    for i in range(N):
        if board[i][i] not in {p, 'T'}:
            g1 = False
        if board[i][N-i-1] not in {p, 'T'}:
            g2 = False
    return g1 or g2

for i in range(T):
    print("Case #{}:".format(i+1),end=' ')
    for j in range(N):
        board[j] = input()
    input()
    if win('X') and not win('O'):
        print("X won")
    elif win('O') and not win('X'):
        print("O won")
    elif going():
        print("Game has not completed")
    else:
        print("Draw")

File: solutions_2453486_1/C++/tfw/Tic-Tac-Tomek.cpp
#include <iostream>
#include <cstdio>
#include <string>

using namespace std;

char board[4][4];

bool gameOver(){
	for(int i = 0; i < 4; i++){
		for(int j = 0; j < 4; j++){
			if (board[i][j] == '.')
				return false;
		}
	}
	return true;
}

bool checkForWin(char player){
	// Horizontal
	for(int i = 0; i < 4; i++){
		bool win = true;
		for(int j = 0; j < 4; j++){
			if (board[i][j] != player && board[i][j] != 'T'){
				win = false;
				break;
			}
		}
		if (win)
			return true;
	}
	
	// Vertical
	for(int i = 0; i < 4; i++){
		bool win = true;
		for(int j = 0; j < 4; j++){
			if (board[j][i] != player && board[j][i] != 'T'){
				win = false;
				break;
			}
		}
		if (win)
			return true;
	}
	
	//Diagonals
	bool win = true;
	for(int i = 0; i < 4; i++){
		if(board[i][i] != player && board[i][i] != 'T'){
			win = false;
			break;
		}
	}
	if(win)
		return true;
	
	win = true;
	for(int i = 0; i < 4; i++){
		if(board[i][3-i] != player && board[i][3-i] != 'T'){
			win = false;
			break;
		}
	}
	if(win)
		return true;
		
	return false;
}

int main(){
	int T; cin >> T;
	for(int i = 0; i < T; i++){
		for(int j = 0; j < 4; j++){
			for(int k = 0; k < 4; k++){
				char c; cin >> c;
				board[j][k] = c;
			}		
		}
		
		if(checkForWin('X')){
			cout << "Case #" << (i+1) << ": X won" << endl;
		} else if (checkForWin('O')){
			cout << "Case #" << (i+1) << ": O won" << endl;
		} else if (!gameOver()){
			cout << "Case #" << (i+1) << ": Game has not completed" << endl;
		} else {
			cout << "Case #" << (i+1) << ": Draw" << endl;
		}
	}
}
